src/qibolab/instruments/icarusqfpga.py

```diff
@@ -1,259 +1,453 @@
+                # This locks the phase of the readout in the demodulation
+                adc = self.ports(pulse.channel).adc
+                start = 0
+
+                end = start + len(i_env)
+                t = np.arange(start, end) / dac_sampling_rate
+                cosalpha = np.cos(
+                    2 * np.pi * pulse.frequency * t + pulse.relative_phase
+                )
+                sinalpha = np.sin(
+                    2 * np.pi * pulse.frequency * t + pulse.relative_phase
+                )
+                wfm = i_env * sinalpha + q_env * cosalpha
+
+                # First we convert the pulse starting time to number of ADC samples
+                # Then, we convert this number to the number of ADC clock cycles (8 samples per clock cycle)
+                # Next, we raise it to the next nearest integer to prevent an overlap between drive and readout pulses
+                # Finally, we ensure that the number is even for the DAC delay conversion
+                delay_start_adc = int(
+                    int(
+                        np.ceil(
+                            self.device.adc_sampling_rate * 1e6 * pulse.start * 1e-9 / 8
+                        )
+                        / 2
+                    )
+                    * 2
+                )
 
-    def stop(self):
-        self._send(b"stop")
+                # For the DAC, currently the sampling rate is 3x higher than the ADC
+                # The number of clock cycles is 16 samples per clock cycle
+                # Hence, we multiply the adc delay clock cycles by 1.5x to align the DAC/ADC pair
+                delay_start_dac = int(delay_start_adc * 1.5)
 
-    def status(self):
-        return self._send(b"status", True)
+                self.device.dac[dac].delay = (
+                    delay_start_dac + self.channel_delay_offset_dac
+                )
+                self.device.adc[adc].delay = (
+                    delay_start_adc + self.channel_delay_offset_adc
+                )
+                # ADC0 complete marks the end of acquisition, so we also need to move ADC0
+                self.device.adc[0].delay = (
+                    delay_start_adc + self.channel_delay_offset_adc
+                )
 
-    def _send(self, payload, retval=False):
-        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
-            s.connect((self.address, self.port))
-            s.sendall(payload)
+                if (
+                    options.acquisition_type is AcquisitionType.DISCRIMINATION
+                    or AcquisitionType.INTEGRATION
+                ):
+                    self.device.program_qunit(
+                        readout_frequency=pulse.frequency,
+                        readout_time=pulse.duration * 1e-9,
+                        qunit=pulse.qubit,
+                    )
+
+            end = start + len(wfm)
+            waveform_array[dac][start:end] += self.device.dac_max_amplitude * wfm
```

Comment:
I know that you initially set to 0, but the idea here is really to set the absolute value, not to sum to a baseline, isn't it?

```suggestion
            waveform_array[dac][start:end] = self.device.dac_max_amplitude * wfm
```

(I'm suggesting to make it more explicit, but you'd even same a bunch of additions...)